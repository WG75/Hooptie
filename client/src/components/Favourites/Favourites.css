.favourites {
    padding: 24px;
    border: 1px solid var(--secondary-color);
    margin: 0 24px;
    flex: 1;
}

.favourites__button {
    display: block;
    margin-left: auto;
}