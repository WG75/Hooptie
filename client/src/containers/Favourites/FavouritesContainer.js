// @flow

import React from 'react';
import { connect } from 'react-redux';
import FavouritesComponent from '../../components/Favourites';
import {
  saveToCollection,
  removeFromCollection,
  searchInCollection,
} from './FavouritesRedux';

type Props = {
  location: Object,
  onSave: (carId: string) => void,
  onRemove: (carId: string) => void,
  searchCollection: (carId: string) => void,
  isSaved: boolean
};

class Favourites extends React.Component<Props, State> {
  componentDidMount() {
    const { location, searchCollection } = this.props;
    const { pathname } = location;
    const carId = pathname.slice(6, 11);

    searchCollection(parseInt(carId, 10));
  }

  render() {
    const {
      onSave, onRemove, isSaved, car,
    } = this.props;
    return (
      <FavouritesComponent
        onSave={() => car && onSave(car.stockNumber)}
        onRemove={() => car && onRemove(car.stockNumber)}
        isSaved={isSaved}
      />
    );
  }
}

function mapStateToProps(state: any) {
  return {
    location: state.router.location,
    isSaved: state.favourites.isSaved,
    car: state.selectedCar.car,
  };
}

function mapDispatchToProps(dispatch: (() => any) => any) {
  return {
    onSave: (carId: string) => dispatch(saveToCollection(carId)),
    onRemove: (carId: string) => dispatch(removeFromCollection(carId)),
    searchCollection: (carId: string) => dispatch(searchInCollection(carId)),
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Favourites);
