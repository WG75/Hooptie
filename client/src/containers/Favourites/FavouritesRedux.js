/* ------------- Actions ------------- */
const ADD_CAR_REQUEST = 'Auto1/Favourites/ADD_CAR_REQUEST';
const REMOVE_CAR_REQUEST = 'Auto1/Favourites/REMOVE_CAR_REQUEST';
const SEARCH_CAR_REQUEST = 'Auto1/Favourites/SEARCH_CAR_REQUEST';

/* ------------- initial state ------------- */
const initialState = {
  collection: [],
  isSaved: false,
};

/* ------------- Reducer ------------- */
export default function reducer(state = initialState, action = {}) {
  switch (action.type) {
    case ADD_CAR_REQUEST:
      return {
        ...state,
        collection: [...state.collection, action.payload.carId],
        isSaved: true,
      };

    case REMOVE_CAR_REQUEST:
      return {
        ...state,
        collection: state.collection.filter(id => id !== action.payload.carId),
        isSaved: false,
      };

    case SEARCH_CAR_REQUEST:
      return {
        ...state,
        error: false,
        loading: false,
        isSaved: state.collection.includes(action.payload.carId),
      };

    default:
      return state;
  }
}

/* ------------- Action Creators ------------- */
export function saveToCollection(carId) {
  return { type: ADD_CAR_REQUEST, payload: { carId } };
}

export function removeFromCollection(carId) {
  return { type: REMOVE_CAR_REQUEST, payload: { carId } };
}

export function searchInCollection(carId) {
  return { type: SEARCH_CAR_REQUEST, payload: { carId } };
}
