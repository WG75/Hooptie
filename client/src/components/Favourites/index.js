// @flow

import React from 'react';
import Button from '../Button';
import './Favourites.css';

type Props = {
  isSaved: boolean,
  onRemove: () => void,
  onSave: () => void
};

const Favourites = ({ isSaved, onRemove, onSave }: Props) => (
  <div className="favourites">
    {isSaved ? (
      <div className="favourites__wrapper">
        <p className="favourites__hint is-regular-medium">
          This car is saved in your collection of favourite items. However if
          you want to remove it click the button below.
        </p>

        <Button onClick={onRemove}>remove</Button>
      </div>
    ) : (
      <div className="favourites__wrapper">
        <p className="favourites__hint is-regular-medium">
          If you like this car, click the button and save it in your collection
          of favourite items.
        </p>

        <Button className="favourites__button" onClick={onSave}>
          save
        </Button>
      </div>
    )}
  </div>
);

export default Favourites;
