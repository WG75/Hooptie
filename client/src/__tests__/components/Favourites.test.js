import React from 'react';
import { create } from 'react-test-renderer';
import Enzyme from '../../config/enzyme';
import Favourites from '../../components/Favourites';

describe('Favourites component', () => {
  test('it matches the snapshot when car is already saved', () => {
    const component = create(<Favourites isSaved />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('it matches the snapshot when car is not saved', () => {
    const component = create(<Favourites isSaved={false} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('it call onSave handler when save bitton is clicked', () => {
    const onSaveMock = jest.fn();
    const wrapper = Enzyme.shallow(
      <Favourites isSaved={false} onSave={onSaveMock} />,
    );

    expect(wrapper.find('Button').length).toBe(1);
    wrapper.find('Button').props().onClick();
    expect(onSaveMock.mock.calls.length).toBe(1);
  });

  it('it call onRemove handler when save bitton is clicked', () => {
    const onRemoveMock = jest.fn();
    const wrapper = Enzyme.shallow(
      <Favourites isSaved onRemove={onRemoveMock} />,
    );

    expect(wrapper.find('Button').length).toBe(1);
    wrapper.find('Button').props().onClick();
    expect(onRemoveMock.mock.calls.length).toBe(1);
  });
});
