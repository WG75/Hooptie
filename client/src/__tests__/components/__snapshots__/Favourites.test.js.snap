// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem component it matches the snapshot when car is already saved 1`] = `
<div
  className="favourites"
>
  <div
    className="favourites__wrapper"
  >
    <p
      className="favourites__hint is-regular-medium"
    >
      This car is saved in your collection of favourite items. However if you want to remove it click the button below.
    </p>
    <button
      className="button button--is_primary    favourites__button"
      disabled={false}
      type="button"
    >
      remove
    </button>
  </div>
</div>
`;

exports[`CartItem component it matches the snapshot when car is not saved 1`] = `
<div
  className="favourites"
>
  <div
    className="favourites__wrapper"
  >
    <p
      className="favourites__hint is-regular-medium"
    >
      If you like this car, click the button and save it in your collection of favourite items.
    </p>
    <button
      className="button button--is_primary    favourites__button"
      disabled={false}
      type="button"
    >
      save
    </button>
  </div>
</div>
`;

exports[`Favourites component it matches the snapshot when car is already saved 1`] = `
<div
  className="favourites"
>
  <div
    className="favourites__wrapper"
  >
    <p
      className="favourites__hint is-regular-medium"
    >
      This car is saved in your collection of favourite items. However if you want to remove it click the button below.
    </p>
    <button
      className="button button--is_primary    favourites__button"
      disabled={false}
      type="button"
    >
      remove
    </button>
  </div>
</div>
`;

exports[`Favourites component it matches the snapshot when car is not saved 1`] = `
<div
  className="favourites"
>
  <div
    className="favourites__wrapper"
  >
    <p
      className="favourites__hint is-regular-medium"
    >
      If you like this car, click the button and save it in your collection of favourite items.
    </p>
    <button
      className="button button--is_primary    favourites__button"
      disabled={false}
      type="button"
    >
      save
    </button>
  </div>
</div>
`;
